homeassistant/util/dt.py

```diff
@@ -337,11 +344,32 @@ def _lower_bound(arr: list[int], cmp: int) -> int | None:
     # For example: if triggering on 2:30 and now is 28.10.2018 2:30 (in DST)
     # we should trigger next on 28.10.2018 2:30 (out of DST), but our
     # algorithm above would produce 29.10.2018 2:30 (out of DST)
-    if tz.datetime_ambiguous(now):
+    if datetime_ambiguous(now):
         check_result = find_next_time_expression_time(
             now + _dst_offset_diff(now), seconds, minutes, hours
         )
-        if tz.datetime_ambiguous(check_result):
+        if datetime_ambiguous(check_result):
             return check_result
 
     return result
+
+
+def datetime_exists(dattim: dt.datetime) -> bool:
+    """Check if a datetime exists."""
+    assert dattim.tzinfo is not None
+    original_tzinfo = dattim.tzinfo
+    # Check if we can round trip to UTC
+    return dattim.replace(tzinfo=None) == dattim.astimezone(UTC).astimezone(
+        original_tzinfo
+    ).replace(tzinfo=None)
+
+
+def datetime_ambiguous(dattim: dt.datetime) -> bool:
+    """Check whether a datetime is ambiguous."""
+    assert dattim.tzinfo is not None
+    non_folded = dattim.replace(fold=0)
+    folded = dattim.replace(fold=1)
+    return not (
+        non_folded.utcoffset() == folded.utcoffset()
+        and non_folded.dst() == folded.dst()
```

Comment:
`non_folded.dst() == folded.dst()` is unnecessary here. Ambiguous datetimes will exist only when the UTC offset has changed — specifically when it has gotten smaller. There should never be a situation where `dt.utcoffset() == dt.replace(fold=not dt.fold).utcoffset()` but `dt.dst() != dt.replace(fold=not dt.fold).utcoffset()` unless someone deliberately designs a `tzinfo` object to behave in this nonsensical way.

Reply:
Removed in a97305e635